### ISSUE TITLE
Dart Core/VM issue: cannot use dart migrate: waiting for isolate to check in ...

### ISSUE THREAD

**Issue by am15h** [dart-lang/sdk#45771]
## Steps to Reproduce

I am trying to migrate my package but  `dart migrate` keeps printing 

```
Attempt:xxx waiting for isolate vm-isolate to check in
Attempt:xxx waiting for isolate dartdev to check in
```
and does not terminate. `flutter doctor -v` and logs attached below.

I also tried to use `Dart SDK version: 2.13.0-93.0.dev (dev) (Mon Mar 1 19:18:22 2021 -0800) on "linux_x64"` as suggested in https://github.com/dart-lang/sdk/issues/45047#issuecomment-791329613, but facing the same problem there too. I can use only `dart --version`  and facing this issue with all the other commands.

## Logs

<details>
<summary>Logs</summary>

```
$ dart migrate .
.
Attempt:11 waiting for isolate vm-isolate to check in
Attempt:11 waiting for isolate dartdev to check in
Attempt:12 waiting for isolate vm-isolate to check in
Attempt:12 waiting for isolate dartdev to check in
Attempt:13 waiting for isolate vm-isolate to check in
Attempt:13 waiting for isolate dartdev to check in
Attempt:14 waiting for isolate vm-isolate to check in
Attempt:14 waiting for isolate dartdev to check in
Attempt:15 waiting for isolate vm-isolate to check in
Attempt:15 waiting for isolate dartdev to check in
Attempt:16 waiting for isolate vm-isolate to check in
Attempt:16 waiting for isolate dartdev to check in
Attempt:17 waiting for isolate vm-isolate to check in
Attempt:17 waiting for isolate dartdev to check in
Attempt:18 waiting for isolate vm-isolate to check in
Attempt:18 waiting for isolate dartdev to check in
Attempt:19 waiting for isolate vm-isolate to check in
```

```
$ flutter doctor -v
Running "flutter pub get" in flutter_tools...                      13.5s
[✓] Flutter (Channel stable, 2.0.5, on Linux, locale en_US.UTF-8)
    • Flutter version 2.0.5 at /home/amish/snap/flutter/common/flutter
    • Framework revision adc687823a (3 days ago), 2021-04-16 09:40:20 -0700
    • Engine revision b09f014e96
    • Dart version 2.12.3

[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at /home/amish/Android/Sdk
    • Platform android-30, build-tools 29.0.2
    • Java binary at: /usr/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_275-8u275-b01-0ubuntu1~18.04-b01)
    • All Android licenses accepted.

[✓] Chrome - develop for the web
    • Chrome at google-chrome

[!] Android Studio (not installed)
    • Android Studio not found; download from https://developer.android.com/studio/index.html
      (or visit https://flutter.dev/docs/get-started/install/linux#android-setup for detailed instructions).

[✓] Connected device (1 available)
    • Chrome (web) • chrome • web-javascript • Google Chrome 86.0.4240.111

! Doctor found issues in 1 category.

```

</details>

**Comment by stereotype441** [comment]
Normally I would classify this as a migration bug, but since the message is comming from deep within the VM (`runtime/vm/dart.cc
`) I suspect it's a VM issue.  @mkustermann any idea what might be going wrong here?

**Comment by am15h** [comment]
Update: I switched to windows and it's running fine on the Dart 2.12.3. Could it be due to some issue with my ubuntu settings?

**Comment by Quijx** [comment]
I too am having this issue. Many `dart` commands do not work but for example `dart --version` does work. Funny enough, while things like `dart -h` or `dart pub -h` do not work, `dart -x` and `dart pub -x` for example to show the help because the flag `-x` is not found.

The exact output of running `dart -h` or other commands is:
```
.
Attempt:11 waiting for isolate vm-isolate to check in
Attempt:11 waiting for isolate dartdev to check in
Attempt:12 waiting for isolate vm-isolate to check in
Attempt:12 waiting for isolate dartdev to check in
Attempt:13 waiting for isolate vm-isolate to check in
Attempt:13 waiting for isolate dartdev to check in
Attempt:14 waiting for isolate vm-isolate to check in
Attempt:14 waiting for isolate dartdev to check in
Attempt:15 waiting for isolate vm-isolate to check in
Attempt:15 waiting for isolate dartdev to check in
Attempt:16 waiting for isolate vm-isolate to check in
Attempt:16 waiting for isolate dartdev to check in
```
and so on. (In case the leading `.` helps debugging)

All flutter commands seem to work.

`dart --version` outputs `Dart SDK version: 2.14.0 (stable) (Mon Sep 6 13:47:59 2021 +0200) on "linux_x64"`

I am also only having this issue on my Kubuntu PC, so yes, it could have something to do with ubuntu. 
And here some system info of mine:
```
Operating System: Kubuntu 20.04
KDE Plasma Version: 5.18.5
KDE Frameworks Version: 5.68.0
Qt Version: 5.12.8
Kernel Version: 5.4.0-84-generic
OS Type: 64-bit
Processors: 4 × Intel® Core™ i5-6300HQ CPU @ 2.30GHz
Memory: 7,7 GiB of RAM
```

**Comment by stereotype441** [comment]
@mraleph any thoughts?

**Comment by mraleph** [comment]
This looks like a startup or shutdown race, we have seen such things before. /cc @a-siva @mkustermann

**Comment by Quijx** [comment]
Is there any workaround I could try?

**Comment by mkustermann** [comment]
@Quijx @am15h  Does this reproduce with newest beta release for you? Do both of you see this on Kubuntu?

```
Attempt:11 waiting for isolate vm-isolate to check in
Attempt:11 waiting for isolate dartdev to check in
```
A VM shutdown happens in stages:
a) disallow isolate(group) creation
b) kill all user isolates & wait until they are dead
c) shutdown VM-internal  isolates & wait until they are dead

From the code it seems that we are in step c) above, i.e. shutting down VM-internal isolates (since we're printing `vm-isolate`). So it probably means that the `dartdev` isolate is considered a system isolate and isn't shutting down.

There were some changes recently (e.g. 3b1e83387fe7e1b03df9259d2dae3f6f8207abfa) that mark some isolates as system isolates, maybe that marks the `dartdev` isolate as system isolate (/cc @bkonyi) ?

From looking at the code, I can see that `dartdev` isn't specially handled - as we do with all other system isolates - during shutdown procedure, see [runtime/vm/dart.cc](https://github.com/dart-lang/sdk/blob/main/runtime/vm/dart.cc#L652):
```c++

  Isolate::KillAllIsolates(Isolate::kInternalKillMsg);  // <-- kills non-system isolates

  // Wait for all isolates, but the service and the vm isolate to shut down.
  if (ServiceIsolate::IsRunning() || KernelIsolate::IsRunning()) {
    ...
    WaitForApplicationIsolateShutdown();  // <-- waits for non-system isolates.
  }

  // Shutdown the kernel isolate.
  ...
  KernelIsolate::Shutdown();

  // Shutdown the service isolate.
  ...
  ServiceIsolate::Shutdown();

  // Wait for the remaining isolate (service isolate) to shutdown.
  WaitForIsolateShutdown();
```

@bkonyi Maybe the `dartdev` isolate has to be explicitly shutdown here?

**Comment by mkustermann** [comment]
> Is there any workaround I could try?

Does passing `--disable-dart-dev` make this issue go away?

**Comment by Quijx** [comment]
@mkustermann The `--disable-dart-dev` does seem to do something. For example `dart --disable-dart-dev -h` does work, but `dart pub --disable-dart-dev get` outputs
```
Running using Flutter version configured in path.


===== CRASH =====
si_signo=Segmentation fault(11), si_code=1, si_addr=(nil)
version=2.14.0 (stable) (Mon Sep 6 13:47:59 2021 +0200) on "linux_x64"
pid=3216, thread=3216, isolate_group=(nil)((nil)), isolate=(nil)((nil))
isolate_instructions=0, vm_instructions=556bf9fcc320
  pc 0x00007f24cb540675 fp 0x00007fff79580430 /lib/x86_64-linux-gnu/libc.so.6+0x18b675
  pc 0x0000556bfa0e8d09 fp 0x00007fff79580460 dart::bin::File::OpenUri(dart::bin::Namespace*, char const*, dart::bin::File::FileOpenMode)+0x19
  pc 0x0000556bf9fa7783 fp 0x00007fff795804f0 dart::bin::DFE::TryReadKernelFile(char const*, unsigned char**, long*)+0x33
  pc 0x0000556bf9fa76de fp 0x00007fff79580530 dart::bin::DFE::ReadScript(char const*, unsigned char**, long*) const+0x2e
  pc 0x0000556bf9fa897c fp 0x00007fff79580600 /home/quijx/.fvm/versions/stable/bin/cache/dart-sdk/bin/dart+0x1cce97c
  pc 0x0000556bf9fa8536 fp 0x00007fff79580680 dart::bin::RunMainIsolate(char const*, char const*, dart::bin::CommandLineOptions*)+0x76
  pc 0x0000556bf9fa9477 fp 0x00007fff795807b0 dart::bin::main(int, char**)+0x607
  pc 0x0000556bf9faa249 fp 0x00007fff795807c0 /home/quijx/.fvm/versions/stable/bin/cache/dart-sdk/bin/dart+0x1cd0249
-- End of DumpStackTrace
/home/quijx/.fvm/versions/stable/bin/internal/shared.sh: Zeile 225:  3216 Abgebrochen             (Speicherabzug geschrieben) "$DART" "$@"
```

On the dart beta channel (`dart --version` outputs `Dart SDK version: 2.15.0-82.1.beta (beta) (Tue Sep 14 08:32:06 2021 +0200) on "linux_x64"`), the problem is still present.

**Comment by bkonyi** [comment]
`--disable-dart-dev` will completely disable the unified CLI, so `dart <command>` won't work in those situations.

The `dartdev` isolate is one of three default system isolates in the standalone VM (the others are the kernel and service isolates). I'll take a look to see if we're missing something in the shutdown path for the `dartdev` isolate.

**Comment by bkonyi** [comment]
I'm unable to reproduce this on `2.14.0`, `2.14.1` and `main`, both on Linux and OSX.

However, it does look like there's an error here in the shutdown logic. System isolates are generally safe to shutdown immediately alongside user/application isolates, with the exception of the kernel, service, and VM isolates. The shutdown logic should be updated to reflect that.

**Comment by bkonyi** [comment]
I've uploaded a speculative [fix here](https://dart-review.googlesource.com/c/sdk/+/213561).

**Comment by Quijx** [comment]
I just randomly tried dart versions 2.10.5 and 2.7.2. The version 2.10.5 does have the problem already, but 2.7.2 works fine. So the problem started somewhere between those versions. Also could it only be in combination with some newer system libraries? Because I would have thought that I used newer versions than 2.10.5 in the past that worked.

**Comment by bkonyi** [comment]
I've landed 58c58af3e8b433715d3339b8ead5406ace482d69 which might resolve this issue. Once we get a dev build with this change, I'll post the version here for people to try.

**Comment by a-siva** [comment]
In any of the error paths in DartDevIsolate::DartDevRunner::RunCallback we do not seem to shutdown the DartDev isolate, if for any reason one of the error paths is taken we could have the main thread start Dart_Cleanup and this would end up in the scenario above were Cleanup is waiting for the DartDev isolate to terminate.
Ensuring that a KillIsolate is done for the DartDev Isolate during Cleanup is not going to cause that isolate to shutdown.
The right fix is to ensure that we shutdown the isolate is all these error paths.
